typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,1002 @@
+
+```typescript
+type A = { x: number };
+type B = { x: number; y: number };
+type C = { y: number };
+
+type Condition<T, V> = T extends V ? number : string;
+
+type N = Condition<B, A>; // type N = number
+type E = Condition<C, A>; // type E = string
+```
+
+Let's move to the Distributive Conditional Types.
+`Exclude` type may looks strange, but what if we will pass union types here?
+It will simply exclude such types in union `T` which exists in union `U` setting `never` in other cases.
+Where `never` type describe the type of values that never present.
+
+```typescript
+type Exclude<T, U> = T extends U ? never : T;
+```
+
+We may think about it as about two nested cycles.
+For each type in `T` and for each type in `U` if `T extends U` return `never` otherwise return `T`.
+
+```typescript
+type A = "a" | "b" | "c";
+type B = Exclude<A, "a" | "b">; // type B = "c"
+type C = Exclude<A, "a" | "b" | "c">; // type C = never
+type D = Exclude<A, "z">; // type D = "a" | "b" | "c"
+```
+
+And using it we can write `Omit` type which will remove properties from type.
+Taking properties of `T` using `keyof T` we exclude from them properties which names are written in `K` and then simply pick remaining ones using `Pick` from the example above.
+
+```typescript
+type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;
+
+type Point = {
+  x: number;
+  y: number;
+  z: number;
+}
+
+type PointX = Omit<Point, "y" | "z">; // type PointX = { x: number }
+```
+
+TypeScript [contains](https://www.typescriptlang.org/docs/handbook/utility-types.html) a lot of predefined conditional and mapped types, so you can use them from the box.
+
+## Going deeper
+
+Build-in concepts are good.
+But as you can see, TypeScript type system is not really a weak one.
+Adding our knowledge from Haskell and type theory we can think about how to get more from this system and what we can build using it.
+This part will show very brief descriptions of different approaches.
+But links, mentioned below, on references and other articles must help you to deeply understand them.
+
+### HKT
+
+You may see that TypeScript type system is not so bad as one may think.
+But it has one important limitation -- absence of kinds.
+Higher-Kinded Types allow us to write types which have their own type constructors as parameters.
+So, using generics instead of abstracting only over types we are able to abstract over types which already abstract over types.
+
+
+Unfortunately we can't write such code and TypeScript type system doesn't allow it.
+```typescript
+interface Collection<F> {
+  create: <A>() => F<A>; // Type 'F' is not generic.
+  insert: <A>(v: A) => (c: F<A>) => F<A>; // Type 'F' is not generic.
+}
+
+const collectionArray: Collection<Array> = {
+  create: () => [],
+  insert: (v) => (c: number[]) => [...c, v]
+};
+```
+
+Fortunately we are able to simulate kinds using defunctionalization which allows us to translate higher-order programs into a first-order language.
```

Comment:
Is there some link to defunctionalization available? https://blog.sigplan.org/2019/12/30/defunctionalization-everybody-does-it-nobody-talks-about-it/ I found smth like this

Reply:
Hm, on wiki https://en.wikipedia.org/wiki/Defunctionalization